codegen/grid.py

```diff
@@ -0,0 +1,1040 @@
+
+
+class SField(Field):
+    """
+    Class to represent stress fields on staggered grid
+    subclass of Field
+    """
+
+    def set(self, dimension, direction):
+        """
+        set number of dimensions and direction of the stress field
+        work out the staggered-ness according to the direction
+        compression stress fields are not staggered
+        sheer stress fields are staggered in the surface normal and force direction
+        stress fields are not staggered in time index
+        i.e. in 3D, field Txx will have staggered = [False, False, False, False], Txy will have staggered = [False, True, True, False]
+        param dimension: number of dimensions, e.g. 3
+        param direction: the direction of the field, e.g. (1,1) for Txx
+        """
+        self.direction = direction
+        staggered = [False] * (dimension+1)
+        if direction[0]==direction[1]:
+            # compression stress, not staggered
+            Field.set(self,dimension,staggered)
+        else:
+            # sheer stress, staggered
+            for i in range(len(direction)):
+                staggered[direction[i]] = True
+            Field.set(self,dimension,staggered)
+
+    def set_free_surface(self, indices, d, b, side):
+        """
+        set free surface boundary condition to boundary d, at index b
+        param indices: list of indices, e.g. [t,x,y,z] for 3D
+        param d: direction of the boundary surface normal
+        param b: location of the boundary (index)
+        side: lower boundary (0) or upper boundary (1)
+        e.g. set_free_surface([t,x,y,z],1,2,0) set y-z plane at x=2 to be lower free surface
+        ghost cells are calculated using reflection of stress fields
+        store the code to populate ghost cells to self.bc
+        """
+        if d not in self.direction:
+            # e.g. x boundary for field Tyy is not required
+            self.bc[d][side] = ''
+            return
+
+        idx = list(indices) # ghost cell
+        idx2 = list(indices) # cell inside domain
+        
+        if not self.staggered[d]:
+            # if not staggered, assign T[d]=0, assign T[d-1]=-T[d+1]
+            idx[d] = b
+            idx2[d] = b
+            eq1 = Eq(self[idx])
+        else:
+            # if staggered, assign T[d-1/2]=T[d+1/2], assign T[d-3/2]=T[d+3/2]
+            idx[d] = b - (1-side)
+            idx2[d] = idx[d] + (-1)**side
+            eq1 = Eq(self[idx], -self[idx2])
+
+        idx[d] -= (-1)**side
+        idx2[d] += (-1)**side
+        eq2 = Eq(self[idx], -self[idx2])
+        self.bc[d][side] = ccode_eq(eq1) +';\n' + ccode_eq(eq2) + ';\n'
+
+class Variable(Symbol):
+    """
+    wrapper class of Symbol to store extra information
+    """
+    def __new__(cls, name, *args):
+        return Symbol.__new__(cls,name)
+
+    def __init__(self, name, value=0, type='int', constant=False):
+        """
+        param type: type string of the variable to be used in generated code
```

Comment:
Don't forget the `:` before `param:` if you want RST style documentation - see e.g. http://thomas-cokelaer.info/tutorials/sphinx/docstring_python.html